### PR TITLE
Linter compatibility

### PR DESCRIPTION
**Pull request by AsaAyers**
This looks like a bigger change than it really is because of having to restore the old `lib/linter.coffee` and `lib/util.coffee`. `linter-eslint` depends on these two files exactly. I'm sure others do too. I think the best way to review this is to review anything after the first two commits.

I have verified this works with `linter-eslint` without any changes.

### CODE REVIEW

File: lib/linter.coffee [hunk 1]
```diff
@@ -1,70 +1,367 @@
...
+fs = require 'fs'
+path = require 'path'
+{CompositeDisposable, Range, Point, BufferedProcess} = require 'atom'
+_ = null
+XRegExp = null
+{MessagePanelView} = require 'atom-message-panel'
 
+# These are NOOPs in linter-plus
+log = -> undefined
+warn = -> undefined
```
Comment (AsaAyers) [inline lib/linter.coffee]: This is the only change I made when I imported this file from AtomLinter/linter.

File: lib/linter.coffee [hunk 2]
```diff
@@ -1,70 +1,367 @@
...
-Panel = require './panel'
-PanelView = require './panel-view'
-Bottom = require './bottom'
-EditorLinter = require './editor-linter'
+fs = require 'fs'
+path = require 'path'
+{CompositeDisposable, Range, Point, BufferedProcess} = require 'atom'
+_ = null
+XRegExp = null
+{MessagePanelView} = require 'atom-message-panel'
```
Comment (steelbrain) [inline lib/linter.coffee]: Ouch

Comment (AsaAyers) [inline lib/linter.coffee]: is "Ouch" supposed to mean something to me here?

**Comment by steelbrain** [comment]
@AsaAyers I am afraid that we cant merge that code here in this repo, All of us had agreed that we are gonna provide the base linter class as an npm module. https://github.com/AtomLinter/atom-linter <-- this is where it'll live

File: lib/linter.coffee (hunk 2 above)
Comment (steelbrain) [inline lib/linter.coffee]: Yeah, I mean, why use atom-message-panel when we already have a panel of ours?

**Comment by AsaAyers** [comment]
but the plugins as they are written TODAY depend on the files this provides.

@hd-deman Is there any point in me even attempting to help with any of this, or are my contributions not welcome here?

**Comment by AsaAyers** [comment]
git isn't handling the rename very gracefully, but otherwise I have provided a PR that makes this backward compatible with the old API so we can just switch linters seamlessly and I'm being told it can't be merged because "we" have decided that all existing linters need to go change their code to remain compatible with the linter they have already been built against.

**Comment by steelbrain** [comment]
@AsaAyers Your contributions are very much welcome. I can't tell you how happy I feel when you contribute here, and I respect the effort that you put in making it work. I get your point now

But if you could just kind-of trim this PR to what is actually required. I mean like remove the usage of `atom-message-panel` and other stuff according to what we already have.

File: lib/linter.coffee (hunk 2 above)
Comment (AsaAyers) [inline lib/linter.coffee]: Do you seriously not understand the idea of compatibility? If this gets merged then you can release the new linter as an update to the old linter with 0 changes to external plugins. You provide a new API and the old API at the same time.

Then you go through and update all the plugins to use the new API and this file eventually dies

**Comment by AsaAyers** [comment]
No, that's not going to happen. That's not making it backward compatible, that's writing new code against the old API. This is just a shim to switch to the new linter without disrupting all of the users and other linter plugins.

**Comment by AsaAyers** [comment]
specirically, I have no way to know what can be trimmed and what can't without testing against all of the other linters which defeats the purpose of doing this.

**Comment by steelbrain** [comment]
What do you mean by backward compatible here specifically, I mean, linter providers don't really style the panel, do they? If not then it shouldn't matter to them how we show it. We could just show it in the new Panel.

File: lib/linter.coffee [hunk 3]
```diff
@@ -1,70 +1,367 @@
...
+    log("beforeSpawnProcess:", command, args, options)
+
+    process = new BufferedProcess({command, args, options,
+                                  stdout, stderr, exit})
+    process.onWillThrowError (err) =>
+      return unless err?
+      if err.error.code is 'ENOENT'
+        ignored = atom.config.get('linter.ignoredLinterErrors')
+        subtle = atom.config.get('linter.subtleLinterErrors')
+        warningMessageTitle = "The linter binary '#{@linterName}' cannot be found."
```
Comment (AsaAyers) [inline lib/linter.coffee]: @steelbrain How should I display this warning if you don't want this to use `MessagePanelView`?

**Comment by hd-deman** [comment]
I want to test it with linter that I contributed to, before merge this. but #22 break it

**Comment by steelbrain** [comment]
I think I am gonna drop another PR and see if it's simpler than this one and provides the same results

**Comment by AsaAyers** [comment]
Why are you allergic to simply importing code that we already know works?

**Comment by steelbrain** [comment]
Because it contains some redundant code which could be removed, less code = easier to maintain

**Comment by AsaAyers** [comment]
https://github.com/AsaAyers/linter-plus/commit/048e46b63f38f0efe237ae09efad77c5b231fec8 <- Look at this commit. That's the only new code, the rest is just importing code that already works in `AtomLinter/linter` with minimal changes

File: lib/linter.coffee [hunk 4]
```diff
@@ -0,0 +1,372 @@
+fs = require 'fs'
```
Comment (park9140) [inline lib/linter.coffee]: This file should just import the npm library @hd-deman created for the core linter, then export it https://github.com/AtomLinter/atom-linter.

File: lib/util.coffee [hunk 5]
```diff
@@ -0,0 +1,42 @@
+path = require 'path'
```
Comment (park9140) [inline lib/util.coffee]: this should be in the core linter library https://github.com/AtomLinter/atom-linter

File: lib/linter.coffee (hunk 3 above)
Comment (steelbrain) [inline lib/linter.coffee]: atom.notifications.addError

File: lib/util.coffee (hunk 5 above)
Comment (AsaAyers) [inline lib/util.coffee]: This an `linter.coffee` are simply copied over from the original linter. They will go away in a few weeks once all plugins have switched over to the new API. Also, there's nothing at that link you sent to use.

File: lib/linter.coffee (hunk 3 above)
Comment (AsaAyers) [inline lib/linter.coffee]: I ended up just simulating an error on the first line of the file. If that doesn't work out I'll try switching it. Thanks

Comment (steelbrain) [inline lib/linter.coffee]: Here's the notification API https://github.com/atom/atom/blob/ce960ff1e1bf0ce9ef3db9bc1fba95e69ee4f003/src/keymap-extensions.coffee#L47

**Comment by AsaAyers** [comment]
I verified using `linter-php` that `lib/linter` will show the error on the first line when configured to show subtle errors.

selection_039

**Comment by steelbrain** [comment]
@AsaAyers The message shown above should be a notification not that kinda message.

**Comment by AsaAyers** [comment]
The first time the error shows up you get this:



After that you can select `Partially` to have it still notify you but to not interrupt you. That looks like this with the old linter:

selection_041

Since you insisted that I can't possibly use the existing UI that used `atom-message-panel`, I changed it to simply move the error to the first line of the file.

**Comment by AsaAyers** [comment]
`atom.notifications.addError` fails at this task as it is a very disruptive notification

**Comment by k2b6s9j** [comment]
Could you change the scope so that it appears as a project error?

**Comment by k2b6s9j** [comment]
Instead of the type being a normal error though, it would be something like a Configuration Error.

**Comment by steelbrain** [comment]
I fail to see or even imagine a notification being a disruptive thing when the `Linter Executable was not found`, that is an error of **critical** severity. and It should be shown as a notification, There is a reason why atom shows notifications of errors instead of just adding it as a decoration to the text editor or adding to the modal.

**Comment by steelbrain** [comment]
screenshot from 2015-06-02 20-39-17

**Comment by AsaAyers** [comment]
> Would you like further notifications to be fully or partially suppressed?

It only shows up there as an error if the user has *explicitly* said that they don't consider it critical.

**Comment by AsaAyers** [comment]
https://github.com/AtomLinter/linter-plus/commit/4564e94f92eb58e6811ce73832337692d9662d8e Fixed, now it looks like your last screenshot

File: package.json [hunk 6]
```diff
@@ -33,5 +33,9 @@
   },
   "devDependencies": {
     "coffeelint": "^1.9.7"
+  },
+  "dependencies": {
+    "lodash": "^2.4.1",
+    "space-pen": "^5.1.1"
```
Comment (steelbrain) [inline package.json]: @AsaAyers Where is this dep being used btw?

File: lib/linter.coffee [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+      args: cmd_list.slice(1)
+    }
+
+  getReportFilePath: (filePath) ->
+    path.join(path.dirname(filePath), @reportFilePath)
+
+  # Private: Provide the node executable path for use when executing a node
+  #          linter
+  getNodeExecutablePath: ->
+    path.join atom.packages.getApmPath(), '..', 'node'
```
Comment (steelbrain) [inline lib/linter.coffee]: Man, I have to admit that this code is extremely un-beautiful. I mean, why do we even need this function, why can't we store the value once and use everywhere?

**Pull request merged**